secrets/azure-keyvault/src/main/kotlin/AzureKeyvaultProvider.kt

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (C) 2024 The ORT Server Authors (See <https://github.com/eclipse-apoapsis/ort-server/blob/main/NOTICE>)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.eclipse.apoapsis.ortserver.secrets.azurekeyvault
+
+import com.azure.security.keyvault.secrets.SecretClient
+
+import org.eclipse.apoapsis.ortserver.secrets.Path
+import org.eclipse.apoapsis.ortserver.secrets.Secret
+import org.eclipse.apoapsis.ortserver.secrets.SecretsProvider
+
+class AzureKeyvaultProvider(private val secretClient: SecretClient) : SecretsProvider {
+    override fun readSecret(path: Path): Secret? =
+        runCatching {
+            secretClient.getSecret(path.path).value?.let { Secret(it) }
+        }.getOrNull()
+
+    override fun writeSecret(path: Path, secret: Secret) {
+        secretClient.setSecret(path.path, secret.value)
+    }
+
+    override fun removeSecret(path: Path) {
+        secretClient.beginDeleteSecret(path.path)
+    }
+
+    override fun createPath(organizationId: Long?, productId: Long?, repositoryId: Long?, secretName: String): Path {
+        val secretType = when {
+            organizationId != null -> "organization-$organizationId"
+            productId != null -> "product-$productId"
+            repositoryId != null -> "repository-$repositoryId"
+            else -> throw IllegalArgumentException(
+                "Either one of organizationId, productId or repositoryId should be specified to create a path."
+            )
+        }
+
+        return Path("$secretType-$secretName")
```

Comment:
Is the dash used here the reason why you do not rely on the default implementation, which uses an underscore?

Reply:
Yes, Azure Key Vault does not allow underscores.

Reply:
As that's quite an interesting restriction, mind adding a code comment for that? Maybe we should even add a `check()` that secret names themselves do not contain dashes either, similar to the [checks for Scaleway](https://github.com/eclipse-apoapsis/ort-server/blob/52e1c7f73b275ae0ccdf1c725a1022d72e092b46/secrets/scaleway/src/main/kotlin/ScalewaySecretsProvider.kt#L218-L228).